### PR TITLE
Improve `deploy.sh`

### PR DESCRIPTION
**Pull request by naszam**
## Rationale

Following #170 pattern I've extended the `deploy.sh` script behaviour to automatically edit `config.sol` by replacing the address with the deployed one (works with both pre-existing `address(0)` or with deployed spell address `address(0x123)`.

Opened for discussion as this could be merged together with #170 to automate further but for now it provides a way to automate further spell crafting process.

A nice add-on would be a commit like pattern as for `dapp-init`.

### CODE REVIEW

**Comment by naszam** [comment]
@brianmcmichael @gbalabasquer I think this PR could be reviewed to be merged on `master` as not pending on #166 discussion, in the meantime.

**Comment by naszam** [comment]
Unified together with #170 to automate further the spell crafting process by relying on Etherscan APIs to retrieve transaction hash.

**Review by godsflaw — CHANGES_REQUESTED** [review]

File: scripts/deploy.sh [hunk 1]
```diff
@@ -4,6 +4,32 @@ set -e
...
+# edit config.sol to add the deployed spell timestamp and block number
+sed -i "s/\($KEY_TIMESTAMP *: *\)[0-9]\+/\1$timestamp/" "$SOURCE"
+sed -i "s/\($KEY_BLOCK *: *\)[0-9]\+/\1$block/" "$SOURCE"
+
+echo -e "Network: $(cast chain)"
+echo "config.sol updated with deployed spell: $spell_address, timestamp: $timestamp and block: $block"
+
+# commit edit change to config.sol
+(set -x; git add src/test/config.sol)
+(set -x; git commit -m "add deployed spell info")
```
[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh:42]: Added tests check in 8581a61

**Review by godsflaw — DISMISSED** [review]

File: scripts/deploy.sh [hunk 2]
```diff
@@ -4,6 +4,39 @@ set -e
...
+KEY_SPELL="deployed_spell"
+KEY_TIMESTAMP="deployed_spell_created"
+KEY_BLOCK="deployed_spell_block"
+
+# Colors
+YELLOW="\033[0;33m"
+PURPLE="\033[0;35m"
+NC="\033[0m"
+
+make && spell_address=$(dapp create DssSpell)
```
[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh:17]: yes, the migration is still in progress, pending some open issue for flattening (see #130)

File: scripts/deploy.sh (hunk 1 above)
[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh]: As discussed in chat I'm testing some improvements to beef the script up with more checks before committing the changes as `make test` is a bit fragile an could return a zero status even if tests doesn't pass.

[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh]: Let me know if a56a15a and b68131f give more assurance, the extra check was inspired by the `dapp upgrade` one (see [here](https://github.com/dapphub/dapptools/blob/master/src/dapp/libexec/dapp/dapp-upgrade#L10))

[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh]: Discussing with @olivdb in chat, I've addressed the request of adding a way to stash and then reload local staging area in 2114dc1 (note that the stash would happen before the script makes edits and then it's reloaded before commiting in case the dev exit the commit edit prompt)

[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh]: after some testing and further discussions, using `trap` we ensure we always exit with restoring the staging area.

File: scripts/deploy.sh [hunk 3]
```diff
@@ -1,9 +1,52 @@
 #!/usr/bin/env bash
 set -e
+trap 'git stash apply' EXIT
```
[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh]: `pop` will only cleanup the stash list, not opposed to use it, done in ad6fcf0

File: scripts/deploy.sh [hunk 4]
```diff
@@ -1,9 +1,52 @@
...
+YELLOW="\033[0;33m"
+PURPLE="\033[0;35m"
+NC="\033[0m"
 
-make && \
-  dapp create DssSpell | \
-  xargs ./scripts/verify.py DssSpell
+# stash any changes in the staging area
+(set -x; git stash push src/test/config.sol)
+echo -e "${YELLOW}Stashing any changes to${NC} ${PURPLE}src/test/config.sol${NC}"
```
[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh:11]: done in c983f2b

File: scripts/deploy.sh [hunk 5]
```diff
@@ -1,9 +1,52 @@
...
+sed -i "s/\($KEY_BLOCK *: *\)[0-9]\+/\1$block/" "$SOURCE"
+
+echo -e "${YELLOW}Network: $(cast chain)${NC}"
+echo -e "${YELLOW}config.sol updated with ${PURPLE}deployed spell:${NC} $spell_address, ${PURPLE}timestamp:${NC} $timestamp and ${PURPLE}block:${NC} $block ${NC}"
+
+make test block="$block" || { echo -e "${PURPLE}Ensure Tests PASS before commiting the config.sol changes${NC}"; exit 1; }
+
+# commit edit change to config.sol
+if [[ $(git status --porcelain src/test/config.sol) ]]; then
+    (set -x; git add src/test/config.sol)
```
[the remark this answers is not in the document]
Comment (naszam) [reply scripts/deploy.sh]: good optimization! done in e8afe75
